core/vm/contracts.go

```diff
@@ -604,25 +614,34 @@ func (c *getValidator) RequiredGas(input []byte) uint64 {
 	return params.GetValidatorGas
 }
 
-// Return the validators that are required to sign this current, possibly unsealed, block. If this block is
+// Return the validators that are required to sign the given, possibly unsealed, block number. If this block is
 // the last in an epoch, note that that may mean one or more of those validators may no longer be elected
 // for subsequent blocks.
+// WARNING: Validator set is always constructed from the canonical chain, therefore this precompile is undefined
+// if the engine is aware of a chain with higher total difficulty.
 func (c *getValidator) Run(input []byte, caller common.Address, evm *EVM, gas uint64) ([]byte, uint64, error) {
 	gas, err := debitRequiredGas(c, input, gas)
 	if err != nil {
 		return nil, gas, err
 	}
 
-	// input is comprised of a single argument:
+	// input is comprised of two arguments:
 	//   index: 32 byte integer representing the index of the validator to get
-	if len(input) < 32 {
+	//   blockNumber: 32 byte integer representing the block number to access
+	if len(input) < 64 {
 		return nil, gas, ErrInputLength
 	}
 
-	index := (&big.Int{}).SetBytes(input[0:32])
+	index := new(big.Int).SetBytes(input[0:32])
 
-	validators := evm.Context.Engine.GetValidators(big.NewInt(evm.Context.BlockNumber.Int64()-1), evm.Context.GetHash(evm.Context.BlockNumber.Uint64()-1))
+	blockNumber := new(big.Int).SetBytes(input[32:64])
+	if blockNumber.Cmp(common.Big0) == 0 || blockNumber.Cmp(evm.Context.BlockNumber) > 0 {
+		return nil, gas, ErrBlockNumberOutOfBounds
+	}
 
+	validators := evm.Context.Engine.GetValidators(new(big.Int).Sub(blockNumber, common.Big1), common.Hash{})
```

Comment:
Mainly because we don't bound the the blockNumber parameter, but also because it isn't actually used for much.

Reply:
Why does not bounding the block number mean we can't include the hash?

Reply:
Because we don't have any way to access arbitrary block hashes here.

Reply:
I've added the method retrieve arbitrary block hashes, and this could be used here, but it isn't necessary so I'd still say we should just pass empty hash. your thoughts?